Improve config location handling

ThoWagen opened the pull request:
This PR makes sure that handlers exists for configs before loading them. Hence, files like "config.ini1", "config.ini.bak" or "config.ini.old" are not loaded instead of "config.ini".

ThoWagen committed:
improve config location handling

EreMaijala commented on module/VuFind/src/VuFind/Config/PathResolver.php:
```diff
@@ -137,6 +139,64 @@ public function getLocalConfigLocation(
+            $itemPath = $path . DIRECTORY_SEPARATOR . $item;
+            $configLocation = $this->getConfigLocationOnPath($itemPath);
+            // ignore locations without a matching handler
+            if ($configLocation === null || $this->configHandlerManager->getForLocation($configLocation) === null) {
```
I think it would be useful to have another method such as `$this->configHandlerManager->hasForLocation($configLocation)` that would just check for the plugin instead of getting it. Even if it's a shared object, getting it does much more than just `->has()`.

ThoWagen replied:
Good Idea! Implemented it.

demiankatz commented on module/VuFind/tests/unit-tests/src/VuFindTest/Config/ConfigManagerTest.php:
```diff
@@ -64,12 +63,10 @@ class ConfigManagerTest extends \PHPUnit\Framework\TestCase
     protected function getConfigManager(): ConfigManager
     {
-        $realResolver = $this->getPathResolver();
+        $config = include APPLICATION_PATH . '/module/VuFind/config/module.config.php';
```
Looks like you could just pass null to `addConfigPluginManagerToContainer` to use this default value, if you want to simplify things.

ThoWagen replied:
The default value was only available in the `addConfigHandlerPluginManagerToContainer` not `addConfigPluginManagerToContainer`, but I added it there now as well.

demiankatz commented on module/VuFind/src/VuFind/Config/PathResolver.php:
```diff
@@ -137,6 +139,64 @@ public function getLocalConfigLocation(
+            // Exclude "." and "..".
+            $ignorePattern = "/^\.{1,2}$/";
+            if (
+                preg_match($ignorePattern, $item)
```
If we're just looking for `.` and `..`, would a simple string comparison be more efficient than a regex? On the other hand, if we want to keep the regex for future flexibility, maybe we should make the pattern a class property for extensibility.

ThoWagen replied:
I think doing simple string comparison for now is simpler. Changed it.

demiankatz commented on module/VuFind/src/VuFind/Config/PathResolver.php:
```diff
@@ -137,6 +139,64 @@ public function getLocalConfigLocation(
+        if ($configNameMatch !== null) {
+            return $configNameMatch;
+        }
+        return null;
```
Why not just:

```suggestion
        return $configNameMatch;
```

ThoWagen replied:
Yeah, that does not make any sense. Changed it.

demiankatz reviewed (changes requested):
Thanks, @ThoWagen -- see below for a few minor suggestions.

ThoWagen committed:
move method to pathresolver

ThoWagen committed:
add hasForLocation method

ThoWagen committed:
small improvements

ThoWagen committed:
add config locations in path cache

ThoWagen commented:
@EreMaijala @demiankatz I also removed the ConfigLocationTrait and added a single execution cache for `getConfigLocationsInPath` as discussed in Slack.

demiankatz reviewed (approved):
Thanks, @ThoWagen and @EreMaijala -- this all looks good to me now! I'm running tests locally and will merge once they pass (which I expect they should).
